### DISCUSSION TITLE
Add a Database Diagnostic Tool

### DISCUSSION THREAD

**Discussion by lat9** [zencart/zencart#6523] [Ideas]
As Zen Cart stores have gotten older, there can be database anomalies that result in unwanted behavior as reported recently on the forum
1. [Products missing descriptions](https://www.zen-cart.com/showthread.php?229928-nightmere-upgrading-Products-not-showing-up-in-their-categories)
2. [Stray entries in the `products_to_categories` table](https://www.zen-cart.com/showthread.php?222983-Image-Handler-5-(for-v1-5-5-v1-5-8)-Support-Thread&p=1401486)

What if we had a Database Diagnostic tool that could detect/report and possibly correct those situations?

I'm thinking of a structure similar to what I've done in [DbIo](https://www.zen-cart.com/downloads.php?do=file&id=2091), where individual tests could be seamlessly added to the overall tool.

**Comment by dbltoe** [comment]
Sounds Great!  What about adding a situation where a previous ID number was reused without removing the old attributes?  I ran into that with one customer. There are tens of thousands of products and a couple hundred reused.  I still occasionally get a log file and have to take it out of the database.

**Reply by lat9** [reply]
I don't understand the situation, previous ID of what?

**Reply by dbltoe** [reply]
product ID

**Reply by drbyte** [reply]
So what's the fix that you do in such cases? Like: what exactly is the problem that you clean up?
Is it that a product wasn't correctly deleted previously, and therefore there are orphaned product_id records in other tables? 
If so, how are you certain that your deletions are actually correct?  Like, how would you know for sure that certain attributes don't actually belong to the re-used product record? ... in an automated way.

Of course, manual human intervention might reveal obvious inconsistencies where a product is associated with attributes wildly unmatching the specified product. 
But to automate that is a different matter.
Even reporting anomalies would be a challenge.

Hence the question: how exactly have you cleaned it up? What did you "take out of the database"? 
And what was in the "log file"?

**Comment by torvista** [comment]
I have code in attribute controller to check
a) all products have a default attribute set
b) all products priced by attribute have a price on all attributes

**Reply by lat9** [reply]
For (b), wouldn't that be dependent on whether the product's base price is 0?

**Reply by torvista** [reply]
yes, it's a specific case example, not a suggestion.

**Comment by scottcwilson** [comment]
Have you looked at Audit? 

https://www.zen-cart.com/downloads.php?do=file&id=2261

**Reply by lat9** [reply]
That's a start, but I'm thinking a bit more separation of the tests and the ability to add checks as we come across additional database anomalies in the future.

**Comment by drbyte** [comment]
Another to check: customers without valid address-book matches, and duplicate address-book entries.
And customer records without corresponding customer_info records (caused by adding customer records through bespoke code instead of core-code methods).

**Comment by lat9** [comment]
Another one: `products_attributes` records which reference non-existent `products`, `products_options` or `products_options_values` entities.

**Comment by gothstone** [comment]
My database goes back to 2005, I would LOVE something like this! I do use audit, and have gone through several rounds of trying to clean my database and undo damage from old plugins and my mistakes, but i know there are still issues.
In particular I would love to see it check changes from previous upgrades. You update from 1.3.7 to 1.5.8, some column didn't change and threw an upgrade exception, yet somehow it still working. Upgrading to 2.0.1 probably didn't fix it cuz the database said it was a 1.5.8 version.
Anything that can improve database integrity would be excellent!
A possible repair direction to take, and kinda what i had to do last time: create a new clean db structure by the default install standards. Then migrate all existing data to that. Problems i have no clue how to automate, old data thats too large for current data type and additional enteries from plugins.
Honestly even a report that called out deviation from the expected schema would be cool. I was comparing schema exports in winmerge. 😞

**Reply by drbyte** [reply]
Ya, like your last comment, I've done that by exporting schema (no data) (to text/.sql file) from current live site, and comparing that with an exported schema (no data) from a fresh new install. 
Exporting both using the same export tool will ensure that the formats are the same, so comparing the text files with WinMerge (or similar tool) is straightforward.

**Comment by drbyte** [comment]
Another check: v2.1.0 retires the use of `banners_on_ssl` and ezpages `page_is_ssl ` fields. Highlighting records where those are set to `0` would expose potential unexpected anomalies where those elements are appearing even though the storeowner had previously set them to 0 to (incorrectly) "hide" them.

**Comment by lat9** [comment]
I'm thinking that the overall tool would be delivered as an encapsulated plugin, with some 'basic' checks.  That 'base' tool would check for other encapsulated plugins providing additional tests in their `admin/includes/classes/DatabaseCheckers` directory.

**Comment by drbyte** [comment]
Another check: look for invalid `configuration_key` entries, such as those with trailing spaces.

eg: https://www.zen-cart.com/showthread.php?36126-Track-Package-from-FedEx-UPS-Postal-DHL-etc&p=1403811#post1403811
